README.md
# USB HID Autofire

[GitHub]([https://github.com/pbek/usb_hid_autofire](https://github.com/hackerman103020/usb_hid_left-right_autoclicker)) |
[Latest release]([https://github.com/pbek/usb_hid_autofire/releases/latest](https://flipc.org/hackerman103020/usb_hid_left-right_autoclicker)) |
[Changelog](CHANGELOG.md) |
[License](LICENSE)

[![Build](https://github.com/pbek/usb_hid_autofire/actions/workflows/build-test.yml/badge.svg)](https://github.com/pbek/usb_hid_autofire/actions/workflows/build-test.yml)

This is a simple Flipper Zero application to send left/right-clicks as a USB HID device.
this is a fork of: https://github.com/pbek/usb_hid_autofire

![Screenshot](flipcorg/banner.png)

## Use-case

- Great for auto-clicking in games like [Idle Wizard](https://store.steampowered.com/app/992070/Idle_Wizard/) 😁

## Installation

Download the [latest release](https://flipc.org/hackerman103020/usb_hid_left-right_autoclicker)
of the *fap* file and put it into the `apps` folder on your SD card of your Flipper Zero. 




CHANGELOG.md
# Changelog

## 0.1
- modded the og repository to be able to switch left and right click 


Keycodes.h
#define KEY_A 0x04 // Keyboard a and A
#define KEY_B 0x05 // Keyboard b and B
#define KEY_C 0x06 // Keyboard c and C
#define KEY_D 0x07 // Keyboard d and D
#define KEY_E 0x08 // Keyboard e and E
#define KEY_F 0x09 // Keyboard f and F
#define KEY_G 0x0a // Keyboard g and G
#define KEY_H 0x0b // Keyboard h and H
#define KEY_I 0x0c // Keyboard i and I
#define KEY_J 0x0d // Keyboard j and J
#define KEY_K 0x0e // Keyboard k and K
#define KEY_L 0x0f // Keyboard l and L
#define KEY_M 0x10 // Keyboard m and M
#define KEY_N 0x11 // Keyboard n and N
#define KEY_O 0x12 // Keyboard o and O
#define KEY_P 0x13 // Keyboard p and P
#define KEY_Q 0x14 // Keyboard q and Q
#define KEY_R 0x15 // Keyboard r and R
#define KEY_S 0x16 // Keyboard s and S
#define KEY_T 0x17 // Keyboard t and T
#define KEY_U 0x18 // Keyboard u and U
#define KEY_V 0x19 // Keyboard v and V
#define KEY_W 0x1a // Keyboard w and W
#define KEY_X 0x1b // Keyboard x and X
#define KEY_Y 0x1c // Keyboard y and Y
#define KEY_Z 0x1d // Keyboard z and Z

#define KEY_1 0x1e // Keyboard 1 and !
#define KEY_2 0x1f // Keyboard 2 and @
#define KEY_3 0x20 // Keyboard 3 and #
#define KEY_4 0x21 // Keyboard 4 and $
#define KEY_5 0x22 // Keyboard 5 and %
#define KEY_6 0x23 // Keyboard 6 and ^
#define KEY_7 0x24 // Keyboard 7 and &
#define KEY_8 0x25 // Keyboard 8 and *
#define KEY_9 0x26 // Keyboard 9 and (
#define KEY_0 0x27 // Keyboard 0 and )


usb_hid_autofire.c
#include <string.h>
#include <furi.h>
#include <furi_hal.h>
#include <furi_hal_usb_hid.h>
#include <gui/gui.h>
#include <input/input.h>
#include <stdio.h>
#include "version.h"
#include "Keycodes.h"

uint8_t Key_code = 0x04;
uint8_t keyCodes[] = {
    0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d,
    0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17,
    0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f, 0x20, 0x21,
    0x22, 0x23, 0x24, 0x25, 0x26, 0x27
};

char All[36] = {'a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z','1','2','3','4','5','6','7','8','9','0'};

// Uncomment to be able to make a screenshot
//#define USB_HID_AUTOFIRE_SCREENSHOT

typedef enum {
  EventTypeInput,
}
EventType;

typedef struct {
  union {
    InputEvent input;
  };
  EventType type;
}
UsbMouseEvent;

bool btn_autofire = false;
float autofire_delay = 1000.0000;
uint32_t cps = 1;
char current1[] = "0";
char * current = "xxx";
char MainSelect[] = "a";
char current3[] = "b";
int prev = 36;
int next = 1;
int minchar = 0;
int maxchar = 36; //length of array
int selected = 0;

static void usb_hid_autofire_render_callback(Canvas * canvas, void * ctx) {
  UNUSED(ctx);
  char autofire_delay_str[12];
  itoa(cps, autofire_delay_str, 10);

  canvas_clear(canvas);

  canvas_set_font(canvas, FontPrimary);
  canvas_draw_str(canvas, 0, 10, "LET/NUM AUTOCLICKER"); //16 charecters long
  canvas_draw_str(canvas, 0, 34, btn_autofire ? "<active>" : "<inactive>");
  canvas_draw_str(canvas, 5, 55, current1);
  canvas_draw_str(canvas, 60, 55, MainSelect);
  canvas_draw_str(canvas, 120, 55, current3);

  canvas_set_font(canvas, FontSecondary);
  canvas_draw_str(canvas, 55, 34, "up/down = time");
  canvas_draw_str(canvas, 0, 22, "press [ok] for on/off");
  canvas_draw_str(canvas, 0, 45, "CPS:");
  canvas_draw_str(canvas, 40, 46, autofire_delay_str);
  canvas_draw_str(canvas, 0, 63, "Left/Right for LET/NUM");
}

static void usb_hid_autofire_input_callback(InputEvent * input_event, void * ctx) {
  FuriMessageQueue * event_queue = ctx;

  UsbMouseEvent event;
  event.type = EventTypeInput;
  event.input = * input_event;
  furi_message_queue_put(event_queue, & event, FuriWaitForever);
}

int32_t usb_hid_autofire_app(void * p) {
  UNUSED(p);
  FuriMessageQueue * event_queue = furi_message_queue_alloc(8, sizeof(UsbMouseEvent));
  furi_check(event_queue);
  ViewPort * view_port = view_port_alloc();

  FuriHalUsbInterface * usb_mode_prev = furi_hal_usb_get_config();
  #ifndef USB_HID_AUTOFIRE_SCREENSHOT
  furi_hal_usb_unlock();
  furi_check(furi_hal_usb_set_config( & usb_hid, NULL) == true);
  #endif

  view_port_draw_callback_set(view_port, usb_hid_autofire_render_callback, NULL);
  view_port_input_callback_set(view_port, usb_hid_autofire_input_callback, event_queue);

  // Open GUI and register view_port
  Gui * gui = furi_record_open(RECORD_GUI);
  gui_add_view_port(gui, view_port, GuiLayerFullscreen);

  UsbMouseEvent event;
  while (1) {
    FuriStatus event_status = furi_message_queue_get(event_queue, & event, autofire_delay);

    if (event_status == FuriStatusOk) {
      if (event.type == EventTypeInput) {
        if (event.input.key == InputKeyBack) {
          break;
        }

        if (event.input.type != InputTypeRelease) {
          continue;
        }

        switch (event.input.key) {
        case InputKeyOk:
btn_autofire = !btn_autofire;
            if (btn_autofire == true) {
                
            }
            else if (btn_autofire == false) {
                
            }

            
          break;
        case InputKeyUp:
            
            if (cps < 100) {
            cps += 1;
          }
           autofire_delay = (1000/cps);
            
          break;
        case InputKeyDown:
          if (cps > 0) {
            cps -= 1;
          }
           autofire_delay = (1000/cps);
          break;
        case InputKeyLeft:

          selected = selected - 1;
          prev = selected;
          next = next - 1;
              //if (selected >= 36) {
               // selected = minchar; 
              //}
              if (selected < 0) {
                selected = 35;
                  prev = 35;
                  next = minchar;
              }
              if (selected == minchar) {
                prev = maxchar; 
              }
              if (selected == 34) {
                next = 35; 
              }
          current1[0]   =  All[prev - 1];
          MainSelect[0] =  All[selected];
          current3[0]   =  All[next];
              if (selected >= 0 && selected <= 35) {
                Key_code = keyCodes[selected];
              } 
            
          break;
        case InputKeyRight:
            
          selected = selected + 1;
          prev = selected;
          next = next + 1;
              if (selected >= 36) {
                selected = minchar; 
              }
              if (selected == minchar) {
                prev = maxchar; 
              }
              if (selected == 35) {
                next = minchar; 
              }
          current1[0]   =  All[prev - 1];
          MainSelect[0] =  All[selected];
          current3[0]   =  All[next];
              if (selected >= 0 && selected <= 35) {
                Key_code = keyCodes[selected];
              } 
            
          break;
        default:
          break;
        }
      }
    }

    if (btn_autofire) {
      furi_hal_hid_kb_press(Key_code);
      // TODO: Don't wait, but use the timer directly to just don't send the release event (see furi_hal_cortex_delay_us)
      furi_hal_hid_kb_release(Key_code);
    }


    view_port_update(view_port);
  }

  furi_hal_usb_set_config(usb_mode_prev, NULL);

  // remove & free all stuff created by app
  gui_remove_view_port(gui, view_port);
  view_port_free(view_port);
  furi_message_queue_free(event_queue);

  return 0;
}
